pkgs/http/lib/src/browser_client.dart

```diff
@@ -52,8 +56,107 @@ class BrowserClient extends BaseClient {
       throw ClientException(
           'HTTP request failed. Client is already closed.', request.url);
     }
+
     var bytes = await request.finalize().toBytes();
+
     var xhr = HttpRequest();
+
+    // Life-cycle tracking is implemented using three completers and the
+    // onReadyStateChange event. The three completers are:
+    //
+    // - connectCompleter (completes when OPENED) (initiates sendingCompleter)
+    // - sendingCompleter (completes when HEADERS_RECEIVED)
+    // (initiates receivingCompleter)
+    // - receivingCompleter (completes when DONE)
+    //
+    // connectCompleter is initiated immediately and on completion initiates
+    // sendingCompleter, and so on.
+    //
+    // Note 'initiated' is not 'initialized' - initiated refers to a timeout
+    // being set on the completer, to ensure the step completes within the
+    // specified timeout.
+    final controller = request.controller;
+
+    if (controller != null) {
+      if (controller.hasLifecycleTimeouts) {
+        // The browser client (which uses XHR) seems not to be able to work with
+        // partial (streamed) requests or responses, so the receive timeout is
+        // handled by the browser client itself.
+        final tracker = controller.track(request, isStreaming: false);
+
+        // Returns a completer for the given state if a timeout is specified
+        // for it, otherwise returns null.
+        Completer<void>? completer(RequestLifecycleState state) =>
+            controller.hasTimeoutForLifecycleState(state)
+                ? Completer<void>()
+                : null;
+
+        final connectCompleter = completer(RequestLifecycleState.connecting);
+        final sendingCompleter = completer(RequestLifecycleState.sending);
+        final receivingCompleter = completer(RequestLifecycleState.receiving);
+
+        // Simply abort the XHR if a timeout or cancellation occurs.
+        void handleCancel(_) => xhr.abort();
+
+        // If a connect timeout is specified, initiate the connectCompleter.
+        if (connectCompleter != null) {
+          unawaited(tracker.trackRequestState(
+            connectCompleter.future,
+            state: RequestLifecycleState.connecting,
+            onCancel: handleCancel,
+          ));
+        }
+
+        xhr.onReadyStateChange.listen((_) {
+          // If the connection is at the OPENED stage and the
+          // connectCompleter has not yet been marked as completed, complete it.
+          if (xhr.readyState == HttpRequest.OPENED) {
+            if (connectCompleter != null) {
+              connectCompleter.complete();
```

Comment:
Does HttpRequest.OPENED really imply that a connection has been established? I thought that it just meant that `.open()` has been called.

Reply:
No, it doesn't ([as suggested here on StackOverflow](https://stackoverflow.com/questions/58324522/how-to-check-that-xmlhttprequest-connection-was-established-before-data-arrive)) so that probably shouldn't be implemented in the BrowserClient.

That said, whilst looking into the browser side of things, I saw that the new fetch API ([caniuse: fetch API](https://caniuse.com/?search=fetch)) supports streams ([caniuse: ReadableStream](https://caniuse.com/?search=ReadableStream)) and could probably be used to implement this functionality as it is on the Dart side.

What kind of support is being aimed for on the browser side? Is there a reason XMLHTTPRequest is still being used or might it be worth looking at moving to the fetch API? - See #595